source/iterator/snapshot.go

```diff
@@ -0,0 +1,322 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/conduitio-labs/conduit-connector-neo4j/config"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j/db"
+	"github.com/neo4j/neo4j-go-driver/v5/neo4j/dbtype"
+)
+
+const (
+	// all Cypher queries used by the [Snapshot] are listed below in the format of Go fmt.
+	getNodeMaxPropertyQueryTemplate         = "MATCH (obj:%s) RETURN obj.%s as %s ORDER BY obj.%s DESC LIMIT 1"
+	getRelationshipMaxPropertyQueryTemplate = "MATCH ()-[obj:%s]-() RETURN obj.%s as %s ORDER BY obj.%s DESC LIMIT 1"
+	getNodesQueryTemplate                   = "MATCH (obj:%s) %s RETURN obj LIMIT %d"
+	getRelationshipsQueryTemplate           = "MATCH (src)-[obj:%s]->(trgt) %s RETURN obj, src, trgt LIMIT %d"
+	opmvLTEWhereClause                      = "obj.%s <= $opmv"
+	opvGTWhereClause                        = "obj.%s > $opv"
+
+	// some helpers for Cypher queries.
+	orderingPropertyMaxValueFieldName = "opmv"
+	orderingPropertyValueFieldName    = "opv"
+	objPlaceholder                    = "obj"
+	srcPlaceholder                    = "src"
+	trgtPlaceholder                   = "trgt"
+
+	// relationship payload-specific fields.
+	sourceNodeField = "sourceNode"
+	targetNodeField = "targetNode"
+)
+
+// Snapshot implements a snapshot logic for the connector.
+type Snapshot struct {
+	driver                   neo4j.DriverWithContext
+	orderingProperty         string
+	keyProperties            []string
+	orderingPropertyMaxValue any
+	entityType               config.EntityType
+	entityLabels             string
+	batchSize                int
+	databaseName             string
+	position                 *position
+	// use a channel here, as it's easier to work with it as a queue
+	records chan map[string]any
+}
+
+// SnapshotParams is incoming params for the [NewSnapshot] function.
+type SnapshotParams struct {
+	Driver           neo4j.DriverWithContext
+	OrderingProperty string
+	KeyProperties    []string
+	EntityType       config.EntityType
+	EntityLabels     []string
+	BatchSize        int
+	DatabaseName     string
+	Position         sdk.Position
+}
+
+// NewSnapshot creates a new instance of the [Snapshot].
+func NewSnapshot(ctx context.Context, params SnapshotParams) (*Snapshot, error) {
+	position, err := parsePosition(params.Position)
+	if err != nil && !errors.Is(err, errNilSDKPosition) {
+		return nil, fmt.Errorf("parse position: %w", err)
+	}
+
+	var (
+		orderingPropertyMaxValue any
+		// join entity labels here to not do this for each individual element
+		entityLabels = strings.Join(params.EntityLabels, ":")
+	)
+	switch pos := position; {
```

Comment:
I think we can use if condition here. Or do we want to expand this switch in future?

Reply:
In this case, we'll need the `else` clause, which is a bit harder to read, in my opinion.

For comparison:
<table>
<tr>
<th><code>if/else</code></th>
<th><code>switch</code></th>
</tr>
<tr>
<td>

```go
if position != nil && position.MaxElement != nil {
	orderingPropertyMaxValue = position.MaxElement
} else {
	orderingPropertyMaxValue, err = getMaxPropertyValue(
		ctx, params.Driver,
		params.DatabaseName, entityLabels, params.OrderingProperty,
		params.EntityType,
	)
	if err != nil && !errors.Is(err, errNoElements) {
		return nil, fmt.Errorf("get ordering property max value: %w", err)
	}
}
```
</td>
<td>

```go
switch pos := position; {
case pos != nil && position.MaxElement != nil:
	orderingPropertyMaxValue = position.MaxElement

default:
	orderingPropertyMaxValue, err = getMaxPropertyValue(
		ctx, params.Driver,
		params.DatabaseName, entityLabels, params.OrderingProperty,
		params.EntityType,
	)
	if err != nil && !errors.Is(err, errNoElements) {
		return nil, fmt.Errorf("get ordering property max value: %w", err)
	}
}
```
</td>
</tr>
</table>

Reply:
ok, I agree with you. Lets leave it